libr/anal/fcn.c

```diff
@@ -48,35 +48,66 @@ R_API const char *r_anal_functiontype_tostring(int type) {
 }
 
 #if READ_AHEAD
-static R_TH_LOCAL ut64 cache_addr = UT64_MAX;
+	#if R2_580
+	// TODO: move into io :?
+	static int read_ahead(RAnal *anal, ut64 addr, ut8 *buf, int len) {
+		static ut64 cache_addr = UT64_MAX;
+		static ut8 cache[1024];
+		const int cache_len = sizeof (cache);
 
-// TODO: move into io :?
-static int read_ahead(RAnal *anal, ut64 addr, ut8 *buf, int len) {
-	static R_TH_LOCAL ut8 cache[1024];
-	const int cache_len = sizeof (cache);
+		if (len < 1) {
+			return 0;
+		}
+		if (len > cache_len) {
+			int a = anal->iob.read_at (anal->iob.io, addr, buf, len); // double read
+			memcpy (cache, buf, cache_len);
+			cache_addr = addr;
+			return a;
+		}
 
-	if (len < 1) {
-		return 0;
-	}
-	if (len > cache_len) {
-		int a = anal->iob.read_at (anal->iob.io, addr, buf, len); // double read
-		memcpy (cache, buf, cache_len);
-		cache_addr = addr;
-		return a;
+		ut64 addr_end = UT64_ADD_OVFCHK (addr, len)? UT64_MAX: addr + len;
+		ut64 cache_addr_end = UT64_ADD_OVFCHK (cache_addr, cache_len)? UT64_MAX: cache_addr + cache_len;
+		bool isCached = ((addr != UT64_MAX) && (addr >= cache_addr) && (addr_end < cache_addr_end));
+		if (isCached) {
+			memcpy (buf, cache + (addr - cache_addr), len);
+		} else {
+			anal->iob.read_at (anal->iob.io, addr, cache, sizeof (cache));
+			memcpy (buf, cache, len);
+			cache_addr = addr;
+		}
+		return len;
 	}
+	#else 
+	static R_TH_LOCAL ut64 cache_addr = UT64_MAX;
```

Comment:
another global that must be moved into ranal

Reply:
Making changes locally. Will rebase once done.